Add test case and documentation for skip_before_filter.

claudiob opened the pull request:
Test case for using skip_before_filter with the options :only and :if
both present. In this case, the :if option will be ignored and :only
will be executed.

---

Closes #14549 since it includes the same commit, rebased on top of the latest master in order to be merged.

laurocaetano committed:
Add test case and documentation for skip_before_filter.

Test case for using skip_before_filter with the options :only and :if
both present. In this case, the :if option will be ignored and :only
will be executed.

Closes #14549 (the commit was cherry-picked from there).

claudiob commented:
This PR simply adds documentation and tests to make the current behavior clearer to the user.

This stems from #9703 where developers expressed their concern about `:only` and `:if` not working well together. This commit clarifies that, when they are used together, `:only` takes priority over `:if`, therefore making `:if` useless.

If Travis is happy, I believe this PR can be merged. 

My only question is whether we should also add tests/docs for the other combination of keywords, to be even clearer:

* `:only` and `:unless`
* `:except` and `:if`
* `:except` and `:unless`

If we do, I can add them in a separate commit or re-write this one, as you prefer.

seuros commented on actionpack/test/controller/filters_test.rb:
```diff
@@ -225,6 +225,18 @@ class ConditionalOptionsSkipFilter < ConditionalFilterController
   end
 
+  class SkipFilterUsingOnlyAndConditional < ConditionalFilterController
+    before_filter :clean_up_tmp
```
this was changed to before_action

claudiob committed:
Add test/doc for :if/:except in skip_before_action

The new test/docs further explain the conflicts that can happen when
mixing `:if`/`:unless` options with `:only`/`:except` options in
`skip_before_action`.

The gist is that "positive" filters always have priority over negative
ones.

The previous commit already showed that `:only` has priority over `:if`.

This commit shows that `:if` has priority over `:except`.

For instance, the following snippets are equivalent:

```ruby
skip_before_action :some_callback, if: -> { condition }, except: action
```

```ruby
skip_before_action :some_callback, if: -> { condition }
```

egilburg commented:
Shouldn't they be used together in an AND-like way?

`only: :index, if: :false` should mean false everywhere, including `:index`

If this doesn't behave that way currently, it feels like a bug to me.

claudiob commented:
@egilburg Yes, I agree with you that the current behavior can be misleading.

This commit simply documents the status quo so developers are aware of it.

Changing the behavior to what you suggest (which might feel more intuitive) would be a backwards-incompatible change, and would need a different type of PR.

laurocaetano commented:
Thanks guys!

 :heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

philipgiuliani commented:
Just came across this issue.

```
skip_before_action :require_user, if: -> { request.format.jpg? }, only: [:show]
```
My temporary fix:
```
skip_before_action :require_user, if: -> { request.format.jpg? && action_name == "show" }
```


Maybe Rails 5 would be a nice chance to fix this issue!

matthewd commented:
@claudiob from a quick reading, this documentation seems to only be true when combining `if`+`only` (etc) for the `skip_*` methods, not for an original `before_action` definition.

The added tests match that theory, but the location of the documentation doesn't seem to.

claudiob commented:
@matthewd I think you are right and… I forgot that this is something we should change for Rails 5! 

Do you think it's too late?

rapcal commented:
> @egilburg Yes, I agree with you that the current behavior can be misleading.
> 
> This commit simply documents the status quo so developers are aware of it.
> 
> Changing the behavior to what you suggest (which might feel more intuitive) would be a backwards-incompatible change, and would need a different type of PR.

9 years later your reply just saved my day :-) Thanks @philipgiuliani
